internal/ingress/controller/parser/kongstate/consumer.go

```diff
@@ -24,6 +24,25 @@ type Consumer struct {
 	K8sKongConsumer configurationv1.KongConsumer
 }
 
+func (c *Consumer) SanitizeCredentials() {
+	redacted := "REDACTED"
+	for _, auth := range c.KeyAuths {
+		auth.Key = &redacted
+	}
+	for _, auth := range c.HMACAuths {
+		auth.Secret = &redacted
+	}
+	for _, auth := range c.JWTAuths {
+		auth.Secret = &redacted
+	}
+	for _, auth := range c.BasicAuths {
+		auth.Password = &redacted
+	}
+	for _, auth := range c.Oauth2Creds {
+		auth.ClientSecret = &redacted
+	}
+}
+
```

Comment:
Two independent thoughts here.

1. #822 made the `parser.Build()` function a pure function (in a sense that it takes an argument, `store.Storer`, and produces a value, `kongstate.KongState` without any side effects other than logging). I believe there are [strong benefits ](http://www.javapractices.com/topic/TopicAction.do?Id=29)(which I'm happy to discuss further in disconnect from this PR, if need be) of making the logic in this part of the codebase functional. I consider the work done in #822 (making `parser.Build` a pure function, and turning all functions that mutate an object into member functions of that object) a step in that direction. Writing new functions is IMO a good opportunity to advance this effort - by writing not mutator functions, but pure functions that return an object if changed.
2. In line with my other comment: this implementation of redaction is subtractive (takes an entire object and wipes away some closed list of fields) which is an anti-pattern for redaction, because addition of a new field and forgetting to update the list is guaranteed to result in a leak.

Therefore, PTAL at #1021 which is an implementation of `kongstate.SanitizedCopy()` which is additive (only copies fields which are known to be safe) and refactors a further part of the codebase in line with the considerations above.

Reply:
What was the purpose of the accompanying `SetCredential()` refactor? I think that may no longer be necessary with the updated strategy in c9ab6de, which drops sensitive items entirely and removes references to them, since it no longer modifies credentials, and it looks like the goal there was to make that modification a function of the credentials rather than the consumers. Assuming so, I think 546a81c to build a copy from chosen fields without that refactor is fine.

I'd originally wanted to produce a redacted version that had credentials and certificates to clearly indicate when the controller wanted to modify them while not disclosing their actual values. Generating appropriate fake values that still exhibited changes when appropriate proved more trouble than it was worth, so I've opted to just drop them entirely.

Reply:
> What was the purpose of the accompanying `SetCredential()` refactor?

Unrelated to the gist of `SanitizedCopy()`. The history behind the `SetCredential()` refactor is the following:
Implementing `SanitizedCopy()` required `KeyAuth`, `HMACAuth`, et al. to implement `SanitizedCopy()`. This was not possible without these types being defined in KIC's codebase. Defining these types created an (unrelated) opportunity to shift the business logic of individual auth types from a "god function" (which `SetCredential` was before the refactor) into specialized "classes".

In short: there's no reason for the `SetCredential()` refactor to be a part of #1021, other than this refactor being a drive-by architecture fix that was easy to carry out given this change. I think that the `SetCredential()` refactor makes sense regardless of the course we take with `SanitizedCopy()`. If you wish, I'm happy to extract the `SetCredential` refactor to a PR unrelated to #1021.

> it looks like the goal there was to make that modification a function of the credentials rather than the consumers

There were two goals:
1. replace mutating an existing `KongState` or any of its subentities with creating a fresh copy (to achieve object immutability)
2. replace "erasing fields known to be secret" (that is, taking KongState as a whole and wiping off cherry-picked fields) with "include only fields known to be non-secret". Slightly different logic, yet better behaved in failure cases.

546a81c does not fulfill the goal number 2.

To summarize: I'd like use to come up with a solution that achieves (1) and (2). #1021 appears to me as such. You seem to be raising some further problems with that solution which are not clear to me.